docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,115 @@
+branch](https://github.com/tendermint/tendermint/tree/tychoish/scratch-node-minimize),
+with the node implementation to see if, without making major
+modifications or changing interfaces, what aspects could be
+simplified. My conclusions, generally: 
+
+- the implementation of the `node` service, doesn't really need the
+  ability to introspect the implementations of reactors in most
+  cases.
+
+- the exception to this, is the cluster of
+  consensus/statesync/blockhain (which depend on eachother,) because
+  `node` itself contains some of their specific initialization logic.
+
+- while the RPC environment construction happens in `node`, this can
+  happen early during `node` construction rather than at
+  initialization time, which reduces the need for the implementation
+  of the node object to track this state.
+
+Ideally, it seems like a node instance should be expressable simply as
+a list of `service.Services` and any knowledge of the internals of any
+of these services could be limited to their constructors. I sketched
+out a generic and simple `groupService` implementation, to handle a
+desecrate stage or sequence of services.  As some services have
+dependencies during construction or initialization and as a starting
+point, a cluster of dependencies can be wrapped as a single
+`service.Service`.
+
+## Goals
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects. 
+  
+  - Reactors should not need to expose their interfaces *within* the
+    implementation of the node type, except in the context of some groups of
+    service. 
+
+- Expose some "unsafe"  way of replacing (some?) existing components, and or
+  inserting arbitrary reactors or services into the tendermint node's
+  initialization.
+  
+  - This may include creating an exported, top-level "unsafe" package with
+    several interfaces and constructors, or simply a function in the
+    node package. 
+
+- Refactor and simplify the process start/initialization logic,
+  separately from--but as a precursor to--permitting user replacement
+  of components.
+
+## Non-Goals
+
+- Fully abstract dependency-based process initialization. 
+
+- Fully abstract replacement of core tendermint components without vendoring.
+
+- Parallelized initialization of components.
+
+## Questions
+
+- To what extent does this new initialization framework need to accommodate
+  the legacy p2p stack? Would it be possible to delay a great deal of this
+  work to the 0.36 cycle to avoid this complexity? 
+  
+- There's clear and common interest in injecting some kind of mempool, and
+  intuitively this makes sense as the highest priority.
+  
+  - Does the existing `mempool.Mempool` interface make sense to expose and
+    commit to, or should we make a smaller interface? 
+    
+  - Is it reasonable for users to bring their own mempool without bringing
```

Comment:
I was for this at first, but the more I read mempool I realized having the data structure be specific to a reactor is the right approach. If a team wants to write a mempool data structure, they should also write the reactor. This would be better than the developer using our reactor and unknown data structure, then coming to ask questions. This way, its a change it out, it's your problem not ours scenario.

Reply:
I think I don't have much of an opinion on this and think we could go either way, with the following notes:

- the `mempool.Mempool` interface is referenced in the `BlockExecutor` in `state`, so we need to find some way for `state` to be able to interface with user-provided memopols. 
-  similarly, this interface gets used by the RPC layer, and I'm not sure how much of that functionality actually depends on the mempool and what we should continue to support.

I'm not really lobbying for either, and it might end up being "both" in a way that's less deslireabe.

Reply:
So the current interface is both for a pluggable mempool data structure into the reactor and to be used in the repo. I would vote for only be used in the repo by other pkgs. The plugability of the mempool DS into the reactor should no be an option imo. The interface it coupled to the implementation anyways as well.

Reply:
> I see from the comment below that we want to support replacing of the mempool reactor only. This makes sense to me as far as a starting point. We can address the support of other reactors in future releases.

agree with this. We should quickly touch base with teams like oasis that replace state sync.

Reply:
They just add a custom state provider because they use a gRPC light client. I talked to them about having a p2p state provider and they said they'd be okay with using that instead

Reply:
> agree with this. We should quickly touch base with teams like oasis that replace state sync.

We have/are, and there's ongoing communication with them.

>  I talked to them about having a p2p state provider and they said they'd be okay with using that instead

Is this work scoped/planned elsewhere that we can reference?

Reply:
> Is this work scoped/planned elsewhere that we can reference?

https://github.com/tendermint/tendermint/issues/6491

I'm looking at getting it in for 0.35 if possible (although I think it could be possible to do in 0.36 and backport to 0.35.1 if it came to it). I just want to get state sync in a "nice" place before I swap out the provider with the p2p one.

Reply:
Totally agree with the conclusions reached in this thread :+1: 

Will think about how to prioritize the P2P state provider for this release

Reply:
I've noted the p2p state provider in my revision, which I think mostly cover this thread.